README.md

```diff
@@ -1,2 +1,221 @@
+```
+
+#### Run the container:
+```
+docker run --rm \
+-e TOWER_ACCESS_TOKEN=$TOWER_ACCESS_TOKEN \
+-e TOWER_API_ENDPOINT=$TOWER_API_ENDPOINT  \
+-e TOWER_PROJECT_NAME=$TOWER_PROJECT_NAME \
+-e WORKFLOW_RUN_ID=$WORKFLOW_RUN_ID \
+ghcr.io/sage-bionetworks-workflows/py-optimize-nextflow:v1.0.0 sh -c \
+'tw --access-token $TOWER_ACCESS_TOKEN --url $TOWER_API_ENDPOINT --output "json" runs view -w $TOWER_PROJECT_NAME -i $WORKFLOW_RUN_ID metrics > metrics.json && python3 optimize-nextflow.py from-json metrics.json'
+```
+
+The output you will receive will look something like:
+
+```
+process {
+
+  withName: synapse_index {
+      maxErrors     = '-1'
+      maxRetries    = 2
+      errorStrategy = { task.attempt <= 2 ? 'retry' : 'finish' }
+
+      cpus   = 2
+      memory = { adj_mem( task, [1.GB] ) }
+  }
+}
+
+def adj_mem(task, progression) {
+    def n_attempts = task.attempt
+    if ( task.exitStatus ) {
+        // Only increase memory if error was memory-related
+        def memory_exit_codes = [104, 134, 137, 139, 143, 247]
+        if ( memory_exit_codes.contains(task.exitStatus) ) {
+            n_attempts = task.attempt
+        } else {
+            n_attempts = task.attempt - 1
+        }
+    }
+
+    if ( n_attempts <= progression.size() ) {
+        return progression[n_attempts - 1]
+    } else {
+        diff = n_attempts - progression.size()
+        return progression.last() * Math.pow(2, diff)
+    }
+}
+```
+
+# Developers
+## Building the docker image for GHCR
+```
+# Step 1: Authenticate with GHCR
+echo $GITHUB_TOKEN | docker login ghcr.io -u USERNAME --password-stdin
+
+# Step 2: Build the Docker image
+docker build -t ghcr.io/sage-bionetworks-workflows/py-optimize-nextflow:latest .
+
+# Step 3: Push the Docker image to GHCR
+docker push ghcr.io/sage-bionetworks-workflows/py-optimize-nextflow:latest
+```
+
+
+# Reducing Compute Costs for Nextflow Pipelines on AWS guidelines
+
+Developing Nextflow pipelines for scientific research and software engineering can be computationally intensive and expensive, especially when running on cloud platforms like AWS. However, there are several strategies you can employ to optimize costs without sacrificing performance.
+
+This article is also avaiable at <https://sagebionetworks.jira.com/wiki/spaces/WF/pages/3518496775/Controlling+Execution+Costs>.
+
+### Utilize Spot Compute Instances
+
+Spot instances are a cost-effective solution for running Nextflow pipelines. These instances can be up to 90% cheaper than on-demand instances. However, there's a tradeoff: spot instances are typically reclaimed after 6-8 hours, which can interrupt long-running tasks. 
+
+**Best Practice:** Use spot instances for short and intermediate steps within your workflow. If a step in your workflow exceeds the typical spot instance duration (6-8 hours), it's better to use on-demand instances for that step. This ensures that you won't lose progress and incur additional costs from having to rerun long tasks.
```

Comment:
[This Best Practice article](https://docs.aws.amazon.com/batch/latest/userguide/best-practices.html#bestpractice5) recommends using on-demand for instances that run >30 mins, so I wonder if we should also have this blurb reflect that stat (which... ugh! This might be trouble for our `RUN_DOCKER` process in the future and might explain a fire I had to put out this morning. Your article is coming in handy right now.)

Reply:
Interesting!  I think we should follow the AWS standard.